### ISSUE TITLE
Implement Peer5 Support

### ISSUE THREAD

**Issue by SpotComms** [lrsjng/h5ai#370]
Peer5 is this nifty service that can be used to speed up file downloads by using a combination of the server and other users to deliver files to the user. Right now it only works with downloads from what I know.

Here is the official documentation of adding to a site, https://peer5.com/downloader/docs.html and if you can't get to the API signup page, use http://peer5.us5.list-manage.com/subscribe/post?u=1a085efa3fd89ffe2532f9e92&amp;id=616a80a156

I tried hardcoding it in for myself, but I'm not much of a web guy

**Comment by shacharz** [comment]
Hi,

The peer5-download attribute needs to be there before dom readystate completes.
In the meanwhile here's a workaround, let me know if you have problems with it:
Add this script after your jquery has loaded.
```
$('body').on('click', '.item.file > a', function(e) {
    if(peer5){
    e.preventDefault();
    var url = e.currentTarget.href;
    peer5.download(url);
   }
});
```

**Comment by xlightwaverx** [comment]
Tried implementing this via h5ai source and rebuilding but the builder calls for the peer5.js script.  I downloaded the peer5.js script and added to /inc/ but the builder is spitting out numerous errors...

<code>[mkr] lint
[fail] jshint - (error) Missing semicolon.   
[fail] jshint - (error) Missing semicolon.   
[fail] jshint - (error) Missing semicolon.   
[fail] jshint - (error) Expected '{' and instead saw 'return'.   
[fail] jshint - (error) Expected '{' and instead saw 'return'.   
[fail] jshint - (error) 'i' is already defined.   
[fail] jshint - (error) Missing semicolon.   
[fail] jshint - (error) Missing semicolon.   
[fail] jshint - (error) Expected '===' and instead saw '=='.   
[fail] jshint - (error) Missing semicolon.   
[fail] jshint - (error) Missing semicolon.   
[fail] jshint - (error) Expected '===' and instead saw '=='.   
[fail] jshint - (error) Expected '{' and instead saw 'return'.   
[fail] jshint - (error) Unexpected use of '++'.   
[fail] jshint - (error) Missing semicolon.   
[fail] jshint - (error) Missing semicolon.   
[fail] jshint - (error) Expected an assignment or function call and instead saw an expression.   
[fail] jshint - (error) Missing semicolon.   
[fail] jshint - (error) A leading decimal point can be confused with a dot: '.2'.   
[fail] jshint - (error) A leading decimal point can be confused with a dot: '.1'.   
[fail] jshint - (error) Expected an assignment or function call and instead saw an expression.   
[fail] jshint - (error) Missing semicolon.   
[fail] jshint - (error) Expected an assignment or function call and instead saw an expression.   
[fail] jshint - (error) Missing semicolon.   
[fail] jshint - (error) Expected an assignment or function call and instead saw an expression.   
[fail] jshint - (error) Missing semicolon.   
[fail] jshint - (error) Expected an assignment or function call and instead saw an expression.   
[fail] jshint - (error) Missing semicolon.   
[fail] jshint - (error) Expected an assignment or function call and instead saw an expression.   
[fail] jshint - (error) Missing semicolon.   
[fail] jshint - (error) Missing semicolon.   
[fail] jshint - (error) Expected an assignment or function call and instead saw an expression.   
[fail] jshint - (error) Missing semicolon.   
[fail] jshint - (error) Expected an assignment or function call and instead saw an expression.   
[fail] jshint - (error) Missing semicolon.   
[fail] jshint - (error) Expected an assignment or function call and instead saw an expression.   
[fail] jshint - (error) Missing semicolon.   
[fail] jshint - (error) Expected an assignment or function call and instead saw an expression.   
[fail] jshint - (error) Missing semicolon.   
[fail] jshint - (error) Expected an assignment or function call and instead saw an expression.   
[fail] jshint - (error) Missing semicolon.   
[fail] jshint - (error) A constructor name should start with an uppercase letter.   
[fail] jshint - (error) A constructor name should start with an uppercase letter.   
[fail] jshint - (error) A constructor name should start with an uppercase letter.   
[fail] jshint - (error) A constructor name should start with an uppercase letter.   
[fail] jshint - (error) A constructor name should start with an uppercase letter.   
[fail] jshint - (error) A constructor name should start with an uppercase letter.   
[fail] jshint - (error) Expected an assignment or function call and instead saw an expression.   
[fail] jshint - (error) Missing semicolon.   
[fail] jshint - (error) Expected '===' and instead saw '=='.   
[fail] jshint - Too many errors. (16% scanned).   

[err] jshint - (error) Missing semicolon.   

———  FILE  /home/buddtest/h5ai/src/_h5ai/client/js/inc/peer5.js  2015-02-17T18:28:08-08:00
001  (function(){function e(){return null}function t(e,t){var i=1024,n=1024*i,r=1024*n,o=1024*r;return e>=0&&i>e?e+" B":e>=i&&n>e?(e/i).toFixed(t)+" KB":e>=n&&r>e?(e/n).toFixed(t)+" MB":e>=r&&o>e?(e/r).toFixe →
002  },receiveDataMessage:function(e,t){this.receiveP2PChunk(e.swarmId,t,e.chunkId,e.payload)},configureBrowserSpecific:function(){window.mozRTCPeerConnection?(this.peerConnectionImpl=r.core.transport.PeerCon →
003  throw e+" is not present in the response headers"},getAllResponseHeaders:function(){return this._responseHeaders},getFileInfo:function(){return this.readyState>=2?this.fileInfo:void 0},abort:function(e){ →
004  },d=function(){try{o.doScroll("left")}catch(e){return setTimeout(d,50),void 0}l("poll")};if("complete"==r.readyState)t.call(e,"lazy");else{if(r.createEventObject&&o.doScroll&&!r.addEventListener){try{n=! →</code>

**Comment by xlightwaverx** [comment]
Managed to find this and do it this way instead.
// @require       http://api.peer5.com/peer5.js?id=z142i5n5qypq4cxr

Still getting these errors as I am unsure how to define it.
<code>
[mkr] lint
[fail] jshint - (error) 'peer5' is not defined.   
[fail] jshint - (error) 'peer5' is not defined.   

[err] jshint - (error) 'peer5' is not defined.   

———  FILE  /home/buddtest/h5ai/src/_h5ai/client/js/scripts.js  2015-02-17T18:44:49-08:00
023  
024      var $ = jQuery;
025  $('body').on('click', '.item.file > a', function(e) {
026      if(peer5){
027      e.preventDefault();
028      var url = e.currentTarget.href;
029      peer5.download(url);
</code>

**Comment by whadar** [comment]
@xlightwaverx  Can you try (window.peer5) instead of (peer5) on line 026?
Also, are you sure the require actually loads the script properly?
If not you can also load the script like this:
```javascript
(function(){document.body.appendChild(document.createElement('script')).src='//api.peer5.com/peer5.js?id=z142i5n5qypq4cxr';})();
```

**Comment by xlightwaverx** [comment]
[err] jshint - (error) 'peer5' is not defined.   

———  FILE  /home/buddtest/h5ai/src/_h5ai/client/js/scripts.js  2015-02-18T09:20:12-08:00
026      if(window.peer5){
027      e.preventDefault();
028      var url = e.currentTarget.href;
029      peer5.download(url);
030     }
031  });
032

**Comment by xlightwaverx** [comment]
Changed 029 to 
window.peer5.download(url);

[mkr] release
[okay] write - /home/buddtest/h5ai/build/h5ai-0.26.1+136~08cd031.zip   

[mkr] successful in 4.404 seconds

**Comment by xlightwaverx** [comment]
Used this method for adding peer5 api support:

(function(){document.body.appendChild(document.createElement('script')).src='//api.peer5.com/peer5.js?id=z142i5n5qypq4cxr';})();

But source code shows no reference to it when viewing it in chrome inspector.

**Comment by xlightwaverx** [comment]
Should I add "http:" to the script call?

**Comment by whadar** [comment]
No, no need for 'http:'  (it will break things when the hosting page is https)
Not sure I understand - you run (function(){document.body.appendChild(document.createElement('script')).src='//api.peer5.com/peer5.js?id=z142i5n5qypq4cxr';})();
but still peer5 is not loaded? 

Do you have it live so I can take a look?

**Comment by xlightwaverx** [comment]
![code](https://cloud.githubusercontent.com/assets/2207951/6252922/b0e3e25c-b76b-11e4-96f4-23f5aa0d4ae9.png)

This is how i am compiling it.

Note: I am not 100% sure if things are where they should be.

**Comment by xlightwaverx** [comment]
http://review.blissroms.com/

The current changes are making it so the script doesn't load, but you can see some source.

**Comment by whadar** [comment]
Yes, the script doesn't load, I thought you're going to execute it after the body was ready.
Right now body has no appendChild therefore it fails.

From what I see, it works if you do 
 $.getScript('//api.peer5.com/peer5.js?id=z142i5n5qypq4cxr')

Instead of (function(){document.body.appendChild(document.createElement('script')).src='//api.peer5.com/peer5.js?id=z142i5n5qypq4cxr';})();

Sorry for the wrong suggestion on the loading...

**Comment by whadar** [comment]
It should work now, let me know if something still fails

**Comment by xlightwaverx** [comment]
I don't know how to do that @whadar.  If I could do that I wouldn't need to post here as @shacharz already suggested that.  

Recompiled and the script is functioning, but peer5 is not working.
http://review.blissroms.com/

**Comment by whadar** [comment]
Looks like peer5 is finally loaded :)

@shacharz and also myself, mistakenly assumed the you're running the snippets at the bottom of the body. Since it's running in the head section, we can do:

```javascript
        //load peer5 with caching
        $.ajax({
            url: '//api.peer5.com/peer5.js?id=z142i5n5qypq4cxr',
            dataType: "script",
            cache:true
        });
        //attach to file items, once the DOM is ready
        $(document).ready(function() {
            $('body').on('click', '.item.file > a', function (e) {
                if (window.peer5) {
                    e.preventDefault();
                    var url = e.currentTarget.href;
                    window.peer5.download(url);
                    return false;
                }
            });
        });
```

Hopefully it's good enough this time :+1:

**Comment by xlightwaverx** [comment]
Thank you gentlemen to all involved.

**Comment by whadar** [comment]
Is it working finally? 
Do you have a working live page?

**Comment by ab835fea** [comment]
@whadar Hello, I tried implementing peer5 for myself as seen here: https://github.com/ab835fea/h5ai/blob/3b8d72e0d813f4dc9b04612c35768c90473c60e6/src/_h5ai/client/js/scripts.js

The script builds fine, however the page on load is blank and this is shown in console: 
https://i.imgur.com/ocd4BKm.png

Any ideas, thanks!

**Comment by whadar** [comment]
Strange, this error doesn't look related directly to peer5... When our code
is commented, do you still get it perhaps?

**Comment by ab835fea** [comment]
@whadar Hi, thanks for the quick reply.

Commenting out the code and re-adding the original string which was removed (see line 59 https://github.com/ab835fea/h5ai/commit/3b8d72e0d813f4dc9b04612c35768c90473c60e6#diff-d6c1f7bb76d0d15f35e6d8ae18f8db7cL38 ) fixes the issue. The page loads fine and with no errors, however its without peer5 obviously.

**Comment by whadar** [comment]
I think the problem is that this part was changed:
    $.ajax({
        url: url,
        data: data,
        type: 'POST',
        dataType: 'json'
    }).done(function (config) {

        modulejs.define('config', config);
        $(function () { modulejs.require(module); });
    });

It's not part of the Peer5 change...

Can you please try this:
https://github.com/whadar/h5ai/commit/c67260744d33729d1ac31c8d7bee273777bcd886

**Comment by ab835fea** [comment]
@whadar 
Hi, I built from your fork.. It built fine and the page loads fine with no errors. However downloads are still served from the server rather than peer5 showing up. In other words peer5 is nowhere to be seen.

**Comment by ab835fea** [comment]
Hi, sorry built from the wrong source. Re-built and it now works fine. Thanks for the help!

**Comment by whadar** [comment]
Awesome! You're welcome!

**Comment by whadar** [comment]
@lrsjng how do you think we should add it to the project in a way that developers can enable/disable it easily?

**Comment by krnlde** [comment]
I like this.

**Comment by whadar** [comment]
@krnlde Great, are you using it? 
Are we ready for a pull request?

**Comment by krnlde** [comment]
@whadar I'm using h5ai but currently not with peer5. But I_will_ use it when stable!

**Comment by lrsjng** [comment]
@whadar hi, the best/preferred way is to wrap it in an extension and put it in the `/src/_h5ai/client/js/inc/ext` folder

**Comment by whadar** [comment]
@lrsjng @krnlde 
I'm without php here and couldn't test, but this should work:
https://github.com/Peer5/h5ai/commit/ac691a8a0acd1691f870ace382fa4ba570d3d59e

Can you please try?
Thanks!

**Comment by whadar** [comment]
not sure I nailed it with the 'enabled' param config though :smile:

**Comment by xlightwaverx** [comment]
I'll just leave this here: https://www.goiww.com/peer5-support-for-incendias-dedicated-servers-and-beyond/

**Comment by shacharz** [comment]
:+1:

**Comment by bmccoy11** [comment]
Here you go. This is using the sample API key, so I'm not sure if this will work or not. If not, just rebuild using your API key.

https://mega.co.nz/#!d0Vm0ZBK!DR1Mb81rGbmhmewSPALqYcu-LBfLcA-XCuyVfvdzKu8

Source: https://github.com/bmccoy11/h5ai-peer5

**Comment by frozeus** [comment]
@whadar works fine and nice integration. maybe you could create a pull create but disable this option as default no ?

**Comment by whadar** [comment]
Here's the pull request, with Peer5 disabled by default:
https://github.com/lrsjng/h5ai/pull/393

**Comment by frozeus** [comment]
Great! :+1:

**Comment by xlightwaverx** [comment]
Woo!

**Comment by lrsjng** [comment]
Added in 0.26.1+149. Thank you!
